### DISCUSSION TITLE
cbs_score_calculator

### DISCUSSION THREAD

**Discussion by SajeevanS** [TheOwenLab/cbs-toolkit#7] [Q&A]
Hey Conor, 

Do you recommend, I give this code a shot, or is it still in the works? 

What do you think? 

Cheers,
Sajeevan

**Comment by conorwild** [comment]
Hey @SajeevanS, give it a shot! At least we will see if it works 🤣, and you can see what the different kinds of data files and scores look like.  
I would say it's plenty good to get you started, but we can explore some other normative datasets / comparisons as things progress.

**Reply by SajeevanS** [reply]
sounds like a good plan. I'll give it a shot and keep you posted. And as you mentioned, its a good starting point and we can discuss other ways to norm the dataset! Thank you once again for all your help Conor!!

**Comment by SajeevanS** [comment]
Interesting, so I managed to compile the code, but got the following error: 

KeyError: "[('spatial_span', 'max_score'), ('spatial_span', 'avg_score'), ('double_trouble', 'final_score'), ('double_trouble', 'pct_CC'), ('double_trouble', 'pct_CI'), ('double_trouble', 'pct_IC'), ('double_trouble', 'pct_II'), ('double_trouble', 'RT_CC'), ('double_trouble', 'RT_CI'), ('double_trouble', 'RT_IC'), ('double_trouble', 'RT_II'), ('feature_match', 'final_score'), ('feature_match', 'attempted'), ('feature_match', 'errors'), ('feature_match', 'max'), ('feature_match', 'correct_score'), ('spatial_planning', 'final_score'), ('polygons', 'final_score'), ('token_search', 'max_score'), ('token_search', 'avg_score')] not in index" 

Any thoughts on this? So for my current study, we are only using the abbreviated cognitive assessments, so 6 out of the 12, could that be causing the issue?

**Reply by conorwild** [reply]
@SajeevanS , could you please copy/paste more of the error including the command as you executed it?

**Reply by SajeevanS** [reply]
Hey, So attached below is my command plus the outputs. Let me know your thoughts: 

MacBook-Pro-102:CBS Sajeevan$ docker run --rm -it -v $PWD:/tmp -w /tmp ghcr.io/theowenlab/cbspython:latest cbs_score_calculator CBS_data.csv my_participants.csv
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
/tmp/CBS_data.csv initially has 1630 rows.
Restructing data frame ...
Clustering scores ...
Parsing Score Features (n_cores=1)
/opt/venv/lib/python3.9/site-packages/cbspython/cbstests.py:267: RuntimeWarning: Mean of empty slice.
  return np.array(item_times).mean()
/opt/venv/lib/python3.9/site-packages/numpy/core/_methods.py:190: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
Parsing Score Features: |██████████████████████████████████████████████████| 100.0% Complete
	Found 251 subjects, and 1629 scores.
Loading /tmp/my_participants.csv...
	Found info on 294 subjects.
Traceback (most recent call last):
  File "/opt/venv/bin/cbs_score_calculator", line 33, in <module>
    sys.exit(load_entry_point('cbspython==0.1.0', 'console_scripts', 'cbs_score_calculator')())
  File "/opt/venv/lib/python3.9/site-packages/cbspython/scripts/score_calculator.py", line 130, in main
    all_scores = all_scores[features_in_norms]
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/frame.py", line 3511, in __getitem__
    indexer = self.columns._get_indexer_strict(key, "columns")[1]
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/multi.py", line 2590, in _get_indexer_strict
    return super()._get_indexer_strict(key, axis_name)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 5782, in _get_indexer_strict
    self._raise_if_missing(keyarr, indexer, axis_name)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/multi.py", line 2610, in _raise_if_missing
    return super()._raise_if_missing(key, indexer, axis_name)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 5845, in _raise_if_missing
    raise KeyError(f"{not_found} not in index")
KeyError: "[('spatial_span', 'max_score'), ('spatial_span', 'avg_score'), ('double_trouble', 'final_score'), ('double_trouble', 'pct_CC'), ('double_trouble', 'pct_CI'), ('double_trouble', 'pct_IC'), ('double_trouble', 'pct_II'), ('double_trouble', 'RT_CC'), ('double_trouble', 'RT_CI'), ('double_trouble', 'RT_IC'), ('double_trouble', 'RT_II'), ('feature_match', 'final_score'), ('feature_match', 'attempted'), ('feature_match', 'errors'), ('feature_match', 'max'), ('feature_match', 'correct_score'), ('spatial_planning', 'final_score'), ('polygons', 'final_score'), ('token_search', 'max_score'), ('token_search', 'avg_score')] not in index"

**Reply by conorwild** [reply]
Thanks @SajeevanS , you are right and this is a bug related to the fact that your trial does not contain all 12 tasks. I am testing a fix, will post an updated image in a few minutes.

**Comment by conorwild** [comment]
Hey @SajeevanS, I have updated the Docker image with a fix. The updated version is still tagged "latest",  so if you re-run your command exactly as you pasted in your last comment you should see that Docker pulls the updated image before running the score calculation. Give it a try!

**Reply by SajeevanS** [reply]
So after running the update script, I get a newer error. 

MacBook-Pro-102:CBS Sajeevan$ docker run --rm -it -v $PWD:/tmp -w /tmp ghcr.io/theowenlab/cbspython:latest cbs_score_calculator CBS_data.csv my_participants.csv
WARNING: The requested image's platform (linux/amd64) does not match the detected host platform (linux/arm64/v8) and no specific platform was requested
/tmp/CBS_data.csv initially has 1630 rows.
Restructing data frame ...
Clustering scores ...
Parsing Score Features (n_cores=1)
/opt/venv/lib/python3.9/site-packages/cbspython/cbstests.py:267: RuntimeWarning: Mean of empty slice.
  return np.array(item_times).mean()
/opt/venv/lib/python3.9/site-packages/numpy/core/_methods.py:190: RuntimeWarning: invalid value encountered in double_scalars
  ret = ret.dtype.type(ret / rcount)
Parsing Score Features: |██████████████████████████████████████████████████| 100.0% Complete
	Found 251 subjects, and 1629 scores.
Loading /tmp/my_participants.csv...
	Found info on 294 subjects.
Creating Matched Norms For Each Subject...
Traceback (most recent call last):
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 3800, in get_loc
    return self._engine.get_loc(casted_key)
  File "pandas/_libs/index.pyx", line 138, in pandas._libs.index.IndexEngine.get_loc
  File "pandas/_libs/index.pyx", line 144, in pandas._libs.index.IndexEngine.get_loc
TypeError: 'user
cfmc0168@researcher-363236.autoregister.com    66.0
cfmc0168@researcher-363236.autoregister.com    66.0
Name: age, dtype: float64' is an invalid key

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/venv/bin/cbs_score_calculator", line 33, in <module>
    sys.exit(load_entry_point('cbspython==0.1.0', 'console_scripts', 'cbs_score_calculator')())
  File "/opt/venv/lib/python3.9/site-packages/cbspython/scripts/score_calculator.py", line 145, in main
    subj_norms = norms.query(
  File "/opt/venv/lib/python3.9/site-packages/cbspython/cbsnorms.py", line 113, in query
    data = self.data_slice(**kwargs).dropna()
  File "/opt/venv/lib/python3.9/site-packages/cbspython/cbsnorms.py", line 74, in data_slice
    data_slice = self.data.loc[tuple(r_index), tuple(c_index)]
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexing.py", line 1068, in __getitem__
    return self._getitem_tuple(key)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexing.py", line 1248, in _getitem_tuple
    return self._getitem_lowerdim(tup)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexing.py", line 942, in _getitem_lowerdim
    return self._getitem_nested_tuple(tup)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexing.py", line 1048, in _getitem_nested_tuple
    obj = getattr(obj, self.name)._getitem_axis(key, axis=axis)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexing.py", line 1306, in _getitem_axis
    locs = labels.get_locs(key)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/multi.py", line 3422, in get_locs
    lvl_indexer = self._get_level_indexer(k, level=i, indexer=indexer)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/multi.py", line 3218, in _get_level_indexer
    start = level_index.get_loc(key.start)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 3807, in get_loc
    self._check_indexing_error(key)
  File "/opt/venv/lib/python3.9/site-packages/pandas/core/indexes/base.py", line 5963, in _check_indexing_error
    raise InvalidIndexError(key)
pandas.errors.InvalidIndexError: user
cfmc0168@researcher-363236.autoregister.com    66.0
cfmc0168@researcher-363236.autoregister.com    66.0
Name: age, dtype: float64

I had some patients who repeated the CBS either twice or more, could this be the issuse? Let me know what you think.

**Reply by conorwild** [reply]
Ah, do you have the same individuals listed twice in `my_participants.csv` ? Double check that there are no duplicated rows...

**Reply by SajeevanS** [reply]
Hey Conor, Yes I do have them duplicated. I wanted to keep the ID the same in both documents, so I didn't remove those. Should I go back and remove all the duplicates?

**Reply by conorwild** [reply]
You should list each participants once in `my_participants.csv`. The program knows how to match each of those entries to all the scores in the CBS data, even if they have taken multiple assessments. So yes, try removing the duplicated participant rows from `my_participants.csv`.

**Reply by SajeevanS** [reply]
It worked Conor! I got 6 csv files that outputted into the same directory. The question I have is, the CBS_data_raw_scores, all of the  are not populated? Is that common, because I was under the impression this would be similar output as the first script that cleans up the data? In the CBS_data_matched_norms, which variables am I looking for in particular? And why are there numerous ages in the age column? I also have some other questions, but if its easier over a quick zoom call. We can find some time to set that up also. 

Cheers,
